### PR TITLE
Make a generic areaDetector for HDF streaming

### LINKED ISSUE #21: Make an object with a current_filename method, for devices to write files in one folder.
We need a way of communicating the scan directory to all detectors so they can all write into it.

We should have an object with stage/unstage method as well as a method 'current_filename', and pass a reference to  all devices that write files.

In future this should be replaced with a service.

### PR DESCRIPTION
**Pull request by coretl**
Also make the same changes to motor that were made to ophyd.v2.epics_demo.Mover.

Fixes #21, #22

### CODE REVIEW

**Comment by coretl** [comment]
We're going to ditch the container build when we move to https://github.com/bluesky/python3-pip-skeleton so ignore that CI failure

File: tests/test_area_detector.py [hunk 1]
```diff
@@ -1,165 +1,214 @@
...
+    assert deta.hdf.name == "deta-hdf"
+
+    for det in (deta, detb):
+        set_sim_value(det.drv.acquire_time, 0.8)
+        set_sim_value(det.drv.image_mode, ImageMode.continuous)
+        set_sim_value(det.hdf.num_capture, 1000)
+        set_sim_value(det.hdf.num_captured, 1)
+        set_sim_value(det.hdf.full_file_name, f"/tmp/123456/{det.name}.h5")
+        set_sim_value(det.drv.array_size_x, 1024)
+        set_sim_value(det.drv.array_size_y, 768)
```
Comment (RAYemelyanova) [inline tests/test_area_detector.py:94]: same here as above. We can replace some of
this, and maybe think about not even setting the readable values:

```python
await det.drv.acquire_time.set(0.8)
await det.drv.image_mode.set(ImageMode.continuous)
await det.hdf.num_capture.set(1000)

# set read only values
set_sim_value(det.hdf.num_captured, 1)
set_sim_value(det.hdf.full_file_name, f"/tmp/123456/{det.name}.h5")
set_sim_value(det.drv.array_size_x, 1024)
set_sim_value(det.drv.array_size_y, 768)
```

I'm not even sure the last two are actually used. If I comment them out the tests still pass.

**Review by RAYemelyanova — CHANGES_REQUESTED** [review]
I have a few questions and comments. I've requested changes on the tests; I think we can change them slightly.

File: src/ophyd_epics_devices/areadetector.py [hunk 2]
```diff
@@ -56,32 +57,41 @@ def __init__(self, prefix: str) -> None:
         self.wait_for_plugins = epics_signal_rw(bool, prefix + "WaitForPlugins")
 
 
-class NDPluginStats(Device):
+class NDPlugin(Device):
+    pass
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_epics_devices/areadetector.py:61]: There will be more NDPlugins in the future, not just NDStats

File: src/ophyd_epics_devices/areadetector.py [hunk 3]
```diff
@@ -112,196 +122,158 @@ def __init__(self, prefix: str) -> None:
...
-        self._compose_datum: Optional[Callable] = None
-        super().__init__(name, config=[self.drv.acquire_time])
+        self.set_readable_signals(config=[self.drv.acquire_time])
+        self.set_name(name)
 
     @AsyncStatus.wrap
     async def stage(self) -> None:
+        # Make a new resource for the new HDF file we're going to open
+        self._resource = _HDFResource()
+        directory = self._namer.need_directory()
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_epics_devices/areadetector.py:203]: Yes, I will strip this back and make a `DirectoryProvider` interface

File: src/ophyd_epics_devices/areadetector.py [hunk 4]
```diff
@@ -112,196 +122,158 @@ def __init__(self, prefix: str) -> None:
...
+    @AsyncStatus.wrap
+    async def trigger(self):
+        await self.drv.image_mode.set(ImageMode.single)
+        frame_timeout = DEFAULT_TIMEOUT + await self.drv.acquire_time.get_value()
+        await self.drv.acquire.set(1, timeout=frame_timeout)
+
+    async def read(self) -> Dict[str, Reading]:
+        # All data goes in the asset docs
+        await self._resource.flush_and_publish(self.hdf)
+        return {}
```
[the remark this answers is not in the document]
Comment (coretl) [reply src/ophyd_epics_devices/areadetector.py:243]: Yes, we originally thought we would need 2 separate Devices, but https://github.com/bluesky/bluesky/issues/1558#issuecomment-1466563991 shows that others are already using Devices that can do both, and it makes the Detector more flexible if it knows how to both

File: tests/test_area_detector.py (hunk 1 above)
Comment (coretl) [reply tests/test_area_detector.py:94]: They come out in the descriptor, but maybe I didn't actually check that in the tests, will add that check too

File: tests/test_area_detector.py [hunk 5]
```diff
@@ -1,165 +1,214 @@
...
+        )
 
-    docs = []
+    assert det.name == "det"
+    assert stats.name == "det-stats"
+    # Set non-default values to check they are set back
+    set_sim_value(det.drv.acquire_time, 0.5)
+    set_sim_value(det.drv.array_counter, 1)
+    set_sim_value(det.drv.image_mode, ImageMode.continuous)
+    set_sim_value(stats.unique_id, 3)
```
[the remark this answers is not in the document]
Comment (RAYemelyanova) [reply tests/test_area_detector.py:37]: Yes please, a comment just explaining this would be useful!

File: tests/test_area_detector.py (hunk 1 above)
Comment (RAYemelyanova) [reply tests/test_area_detector.py:94]: yep adding this check sounds like a good idea. Am happy for you to leave these as set_sim_value with an appropriate comment as you've explained above :+1:

**Review by RAYemelyanova — APPROVED** [review]
looks good!

**Pull request merged**